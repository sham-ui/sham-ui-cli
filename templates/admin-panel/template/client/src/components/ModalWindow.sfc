<template>
    {% import CubeAnimation from 'sham-ui-cube-animation' %}

    <div class="modal" data-test-modal>
        {% if !dataSaving %}
            <span :onclick=\{{::this._closeClick}} class="close" title="Close Modal" data-test-close-button>&times;</span>
        {% endif %}
        <div class="modal-content">
            <div class="container">
                <h4>\{{title}}</h4>
                <p class="content">\{{text}}</p>

                <div class="footer">
                    {% if dataSaving %}
                        <CubeAnimation />
                    {% else %}
                        <button :onclick=\{{::this._okClick}} data-test-ok-button class=\{{isDanger ? "btn-danger" : ""}}>
                            \{{okText}}
                        </button>
                        <button class="btn-secondary" :onclick=\{{::this._closeClick}} data-test-cancel-button>
                            \{{cancelText}}
                        </button>
                    {% endif %}
                </div>
            </div>
        </div>
    </div>
</template>

<script>
    import { mergeWithTemplate, options } from 'sham-ui-macro/babel.macro';

    @mergeWithTemplate
    class ModalWindow extends Template {
        @options title = '';
        @options text = '';
        @options cancelText = 'Cancel';
        @options okText = 'Ok';
        @options isDanger = false;
        @options dataSaving = false;

        @options onOk() {}
        @options onClose() {}

        _enableSaving() {
            this.update( {
                dataSaving: true
            } );
        }

        _okClick() {
            this._enableSaving();
            this.options.onOk();
        }

        _closeClick() {
            this._enableSaving();
            this.options.onClose();
        }
    }

    export default ModalWindow;
</script>
