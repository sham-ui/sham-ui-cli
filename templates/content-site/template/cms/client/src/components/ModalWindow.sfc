<template>
    {% import CubeAnimation from 'sham-ui-cube-animation' %}

    <div class="modal" data-test-modal>
        {% if !dataSaving %}
            <span :onclick=\{{this$._closeClick}} class="close" title="Close Modal" data-test-close-button>&times;</span>
        {% endif %}
        <div class="modal-content">
            <div class="container">
                <h4>\{{title}}</h4>
                <p class="content">\{{text}}</p>

                <div class="footer">
                    {% if dataSaving %}
                        <CubeAnimation />
                    {% else %}
                        <button :onclick=\{{this$._okClick}} data-test-ok-button class=\{{isDanger ? "btn-danger" : ""}}>
                            \{{okText}}
                        </button>
                        <button class="btn-secondary" :onclick=\{{this$._closeClick}} data-test-cancel-button>
                            \{{cancelText}}
                        </button>
                    {% endif %}
                </div>
            </div>
        </div>
    </div>
</template>

<script>
    function ModalWindow( options, update ) {
        const dataSaving = $();
        const onOk = $();
        const onClose = $();

        const state = options( {
            [ $.title ]: '',
            [ $.text ]: '',
            [ $.cancelText ]: 'Cancel',
            [ $.okText ]: 'Ok',
            [ $.isDanger ]: false,
            [ dataSaving ]: false,
            [ onOk ]() {},
            [ onClose ]() {}
        } );

        const enableSaving = () => update( {
            [ dataSaving ]: true
        } );

        this$._okClick = () => {
            enableSaving();
            state[ onOk ]();
        };

        this$._closeClick = () => {
            enableSaving();
            state[ onClose ]();
        };
    }

    export default Component( Template, ModalWindow );
</script>
