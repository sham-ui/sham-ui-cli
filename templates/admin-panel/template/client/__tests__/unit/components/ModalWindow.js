import * as directives from 'sham-ui-directives';
import ModalWindow  from '../../../src/components/ModalWindow.sfc';
import renderer from 'sham-ui-test-helpers';

it( 'renders correctly', () => {
    const meta = renderer( ModalWindow, {
        directives
    } );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'close', async() => {
    expect.assertions( 1 );
    const onClose = jest.fn();
    const meta = renderer( ModalWindow, {
        directives,
        onClose
    } );
    meta.component.container.querySelector( '[data-test-close-button]' ).click();
    await new Promise( resolve => setImmediate( resolve ) );
    expect( onClose ).toHaveBeenCalledTimes( 1 );
} );


it( 'cancel', async() => {
    expect.assertions( 1 );
    const onClose = jest.fn();
    const meta = renderer( ModalWindow, {
        directives,
        onClose
    } );
    meta.component.container.querySelector( '[data-test-cancel-button]' ).click();
    await new Promise( resolve => setImmediate( resolve ) );
    expect( onClose ).toHaveBeenCalledTimes( 1 );
} );


it( 'ok', async() => {
    expect.assertions( 1 );
    const onOk = jest.fn();
    const meta = renderer( ModalWindow, {
        directives,
        onOk
    } );
    meta.component.container.querySelector( '[data-test-ok-button]' ).click();
    await new Promise( resolve => setImmediate( resolve ) );
    expect( onOk ).toHaveBeenCalledTimes( 1 );
} );
