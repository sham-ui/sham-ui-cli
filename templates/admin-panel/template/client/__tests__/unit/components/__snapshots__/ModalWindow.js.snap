// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Object {
  "Constructor": "ModalWindow",
  "Options": Object {
    "ID": "component",
    "cancelText": "Cancel",
    "closeIcon": "×",
    "dataSaving": false,
    "directives": Object {
      "EventListener": [Function],
      "disabled": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
      "ref": [Function],
    },
    "isDanger": false,
    "okText": "Ok",
    "onClose": [Function],
    "onOk": [Function],
    "text": "",
    "title": "",
  },
  "html": "
<div class=\\"modal\\" data-test-modal=\\"\\"><span class=\\"close\\" title=\\"Close Modal\\" data-test-close-button=\\"\\">×</span>
  <!--if-->
  <div class=\\"modal-content\\">
    <div class=\\"container\\">
      <h4></h4>
      <p class=\\"content\\"></p>
      <div class=\\"footer\\">
        <button data-test-ok-button=\\"\\" class=\\"\\">Ok</button>
        <button class=\\"btn-secondary\\" data-test-cancel-button=\\"\\">Cancel</button>
      </div>
    </div>
  </div>
</div>
",
}
`;
